trieste/utils/pareto.py

```diff
@@ -101,10 +114,168 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
+            ],
+            axis=0,
+        )
+
+        divide_conquer_cells = tf.stack(
+            [
+                tf.zeros(number_of_objectives, dtype=tf.int32),
+                (int(pseudo_front_idx.shape[0]) - 1)
+                * tf.ones(number_of_objectives, dtype=tf.int32),
+            ],
+            axis=0,
+        )[None]
+
+        total_size = tf.reduce_prod(max_front - min_front)
+
+        def while_body(
+            divide_conquer_cells: TensorType,
+            lower_result: TensorType,
+            upper_result: TensorType,
+        ) -> Tuple[TensorType, TensorType, TensorType]:
+            divide_conquer_cells_unstacked = tf.unstack(divide_conquer_cells, axis=0)
+            cell = divide_conquer_cells_unstacked[-1]
+            divide_conquer_cells_new = tf.cond(
+                tf.not_equal(tf.size(divide_conquer_cells_unstacked[:-1]), 0),
+                lambda: tf.stack(divide_conquer_cells_unstacked[:-1]),
+                lambda: tf.zeros([0, 2, number_of_objectives], dtype=tf.int32),
+            )
+
+            arr = tf.range(number_of_objectives)
+            lower_idx = tf.gather_nd(pseudo_front_idx, tf.stack((cell[0], arr), -1))
+            upper_idx = tf.gather_nd(pseudo_front_idx, tf.stack((cell[1], arr), -1))
+            lower = tf.gather_nd(pseudo_front, tf.stack((lower_idx, arr), -1))
+            upper = tf.gather_nd(pseudo_front, tf.stack((upper_idx, arr), -1))
+
+            test_accepted = self._is_test_required((upper - jitter) < front)
+            lower_result_final, upper_result_final = tf.cond(
+                test_accepted,
+                lambda: self._accepted_test_body(lower_result, upper_result, lower_idx, upper_idx),
+                lambda: (lower_result, upper_result),
+            )
+
+            test_rejected = self._is_test_required((lower + jitter) < front)
+            divide_conquer_cells_final = tf.cond(
+                tf.logical_and(test_rejected, tf.logical_not(test_accepted)),
+                lambda: self._rejected_test_body(
+                    cell, lower, upper, divide_conquer_cells_new, total_size, threshold
+                ),
+                lambda: divide_conquer_cells_new,
+            )
+
+            return divide_conquer_cells_final, lower_result_final, upper_result_final
+
+        _, lower_result_final, upper_result_final = tf.while_loop(
+            lambda divide_conquer_cells, lower_result, upper_result: len(divide_conquer_cells) > 0,
+            while_body,
+            loop_vars=[divide_conquer_cells, lower_result, upper_result],
+            shape_invariants=[
+                tf.TensorShape([None, 2, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+                tf.TensorShape([None, number_of_objectives]),
+            ],
+        )
+        return BoundedVolumes(lower_result_final, upper_result_final)
+
+    def _accepted_test_body(
```

Comment:
The following few methods could all be marked `@classmethod` or `@staticmethod`. Also might make sense to move _is_test_required down here with the other releated methods.